src/Elm/Kernel/Platform.js

```diff
@@ -15,43 +15,177 @@ import Result exposing (isOk)
 // PROGRAMS
 
 
-var _Platform_worker = F4(function(impl, flagDecoder, debugMetadata, args)
+var _Platform_worker = F3(function(impl, flagDecoder, debugMetadata)
 {
-	return _Platform_initialize(
-		flagDecoder,
-		args,
-		impl.__$init,
-		impl.__$update,
-		impl.__$subscriptions,
-		function() { return function() {} }
-	);
+	var init = function(args)
+	{
+		return _Platform_initialize(
+			flagDecoder,
+			args,
+			null,
+			null,
+			null,
+			function() { return function() {} },
+			impl
+		);
+	};
+
+	/**__DEBUG/
+	init.hotReloadData = {
+		__$impl: impl,
+		__$platform_effectManagers: _Platform_effectManagers,
+		__$scheduler_enqueue: __Scheduler_enqueue
+	};
+	//*/
+
+	return init;
 });
 
 
 
 // INITIALIZE A PROGRAM
 
 
-function _Platform_initialize(flagDecoder, args, init, update, subscriptions, stepperBuilder)
+function _Platform_initialize(flagDecoder, args, _init, _update, _subscriptions, stepperBuilder, impl)
 {
+	// Old versions of elm/browser do not send the `impl` – instead, they send parts of it separately.
+	// Sending the whole object is required for hot reloading.
+	if (!impl)
+	{
+		impl = {
+			__$init: _init,
+			__$update: _update,
+			__$subscriptions: _subscriptions
+		}
+	}
+
 	var result = A2(__Json_run, flagDecoder, __Json_wrap(args ? args['flags'] : undefined));
 	__Result_isOk(result) || __Debug_crash(2 /**__DEBUG/, __Json_errorToString(result.a) /**/);
 	var managers = {};
-	var initPair = init(result.a);
+	var initPair = impl.__$init(result.a);
```

Comment:
You’ll see this a lot in the diff: Instead of popping off stuff from from `impl` (like `var init = imp.__$init; init()`), we pass around the whole `impl` object and read from it all the time (`impl.__$init()`).

This allows us to simply mutate the `impl` object later (`impl.__$init = newInit`) to swap in new versions of the functions. That’s the core of how the hot reloading works!